heartbeat/monitors/active/tcp/endpoint.go

```diff
@@ -0,0 +1,91 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package tcp
+
+import (
+	"fmt"
+	"net"
+	"net/url"
+	"strconv"
+	"strings"
+)
+
+// endpoint configures a host with all port numbers to be monitored by a dialer
+// based job.
+type endpoint struct {
+	Scheme   string
+	Hostname string
+	Ports    []uint16
+}
+
+// perPortURLs returns a list containing one URL per port
+func (e endpoint) perPortURLs() (urls []*url.URL) {
+	for _, port := range e.Ports {
+		urls = append(urls, &url.URL{
+			Scheme: e.Scheme,
+			Host:   net.JoinHostPort(e.Hostname, strconv.Itoa(int(port))),
+		})
+	}
+
+	return urls
+}
+
+// makeEndpoints creates a single endpoint struct for each host/port permutation.
+// Set `defaultScheme` to choose which scheme is used if not explicit in the host config.
+func makeEndpoints(hosts []string, ports []uint16, defaultScheme string) (endpoints []endpoint, err error) {
+	for _, h := range hosts {
+		scheme := defaultScheme
+		host := ""
+		u, err := url.Parse(h)
+
+		if err != nil || u.Host == "" {
+			host = h
+		} else {
+			scheme = u.Scheme
+			host = u.Host
+		}
+		debugf("Add tcp endpoint '%v://%v'.", scheme, host)
+
+		switch scheme {
+		case "tcp", "plain", "tls", "ssl":
+		default:
+			err := fmt.Errorf("'%v' is not a supported connection scheme in '%v'", scheme, h)
+			return nil, err
+		}
+
+		pair := strings.SplitN(host, ":", 2)
```

Comment:
is IPv6 supported? This would break it. Consider using SplitHostPort: https://golang.org/pkg/net/#SplitHostPort

Reply:
I just pushed a fix here where I rewrote all this logic and added a ton of tests as well. It's a weird thing to parse since we support hosts, ips, and URLs (there isn't an RFC for this either AFAICT), but I think then new logic does what the docs say, and has tests to back it up (including ipv6 tests).